### ISSUE TITLE
registerPlugin() will break UndoManger in Collaboration Extension

### ISSUE THREAD

**Issue by Alecyrus** [ueberdosis/tiptap#2761]
### What’s the bug you are facing?

`Collaboration` Extension cannot not work well with `BubbleMenu`, which is registered by using `registerPlugin()` after editori s initialized.

Specifically, once you call `registerPlugin()` after editor is initialized, `UndoManger` does not work.
```javascript

import { Editor, EditorContent } from '@tiptap/vue-3';
import Collaboration from '@tiptap/extension-collaboration';
import StarterKit from '@tiptap/starter-kit';
import CollaborationCursor from '@tiptap/extension-collaboration-cursor';
import * as Y from 'yjs';
import { WebrtcProvider } from 'y-webrtc';
import Placeholder from '@tiptap/extension-placeholder';
import { Plugin } from 'prosemirror-state';

const editor = shallowRef<Editor>();

const ydoc = new Y.Doc();

const provider = new WebrtcProvider(
  'tiptap-collaboration-cursor-extension',
  ydoc,
);

editor.value = new Editor({
  extensions: [
    StarterKit.configure({
      // The Collaboration extension comes with its own history handling
      history: false,
    }),
    Collaboration.configure({
      document: ydoc
    }),
    CollaborationCursor.configure({
      provider: provider,
      user: {
        name: 'Cyndi Lauper',
        color: '#f783ac',
      },
    }),
    Placeholder.configure({
      placeholder:
        'Write something … It’ll be shared with everyone else looking at this example.',
    }),
  ],
});
// comment this line to make undoManager work
editor.value.registerPlugin(new Plugin({}));
</script>
```

If you comment last line, `UndoManger` is back to work.

### Can you provide a CodeSandbox?

[Reproduction Link](https://codesandbox.io/s/kind-dew-urdc6t?file=/src/App.vue)

### Did you update your dependencies?

- [X] Yes, I’ve updated my dependencies to use the latest version of all packages.

### Are you sponsoring us?

- [X] Yes, I’m a sponsor. 💖

**Comment by Alecyrus** [comment]
@hanspagel  @philippkuehn  🙏🙏🙏

**Comment by Alecyrus** [comment]
https://github.com/yjs/y-prosemirror/issues/114

**Comment by Alecyrus** [comment]
There is a workaround (make sure yUndoPlugin() is the last plugin registered ):

```javascript
// collaboration.ts
addProseMirrorPlugins() {
    const fragment = this.options.fragment
      ? this.options.fragment
      : this.options.document.getXmlFragment(this.options.field);
 
    // delete yUndoPlugin()
    // return [ySyncPlugin(fragment), yUndoPlugin()]; 
    retuern [ySyncPlugin(fragment)]
  },
```
and register `yUndoPlugin()` manually after all other plugins are registered. 
```javascript 
editor.registerPlugin(yUndoPlugin());
```

**Comment by ruslanchek** [comment]
> There is a workaround (make sure yUndoPlugin() is the last plugin registered ):
> 
> ```js
> // collaboration.ts
> addProseMirrorPlugins() {
>     const fragment = this.options.fragment
>       ? this.options.fragment
>       : this.options.document.getXmlFragment(this.options.field);
>  
>     // delete yUndoPlugin()
>     // return [ySyncPlugin(fragment), yUndoPlugin()]; 
>     retuern [ySyncPlugin(fragment)]
>   },
> ```
> 
> and register `yUndoPlugin()` manually after all other plugins are registered.
> 
> ```js
> // in onCreate() Hook
> editor.registerPlugin(yUndoPlugin());
> ```

What file is this? I mean collaboration.ts is the plugin's file?

**Comment by Alecyrus** [comment]
Yeah.... You need to fork the code of `@tiptap/extension-collaboration`.

**Comment by ruslanchek** [comment]
> Yeah.... You need to fork the code of `@tiptap/extension-collaboration`.

Ok i will try. Thank you. If this will fix that, I'll make a PR

**Comment by Alecyrus** [comment]
@ruslanchek  I prefer the upstream library to fix this problem.

**Comment by ryanb** [comment]
What worked for me is re-register the plugin after all other plugins are registered.

```
editor.unregisterPlugin(yUndoPluginKey);
editor.registerPlugin(yUndoPlugin());
```

This means forking the collaboration plugin isn't necessary. I haven't done any benchmarking to see if registering the y-undo plugin twice impacts performance, but at least it works as a quick fix until a more permanent solution is implemented.

**Comment by Alecyrus** [comment]
I maintain all extensions of my own to gain more flexibility for my project,  after all, the code for these extensions is very simple and easy to modify.

**Comment by hamflx** [comment]
I had the same problem and this is how I fixed it:

```javascript
import { Collaboration } from '@tiptap/extension-collaboration'
import { ySyncPlugin, yUndoPlugin, yUndoPluginKey } from 'y-prosemirror'

export const CollaborationExtension = Collaboration.extend({
  addProseMirrorPlugins () {
    const fragment = this.options.fragment
      ? this.options.fragment
      : this.options.document.getXmlFragment(this.options.field)

    const yUndoPluginInstance = yUndoPlugin()
    const originalUndoPluginView = yUndoPluginInstance.spec.view
    yUndoPluginInstance.spec.view = view => {
      const undoManager = yUndoPluginKey.getState(view.state).undoManager
      if (undoManager.restore) {
        undoManager.restore()
        undoManager.restore = () => {}
      }
      const viewRet = originalUndoPluginView(view)
      return {
        destroy: () => {
          const hasUndoManSelf = undoManager.trackedOrigins.has(undoManager)
          const observers = undoManager._observers
          undoManager.restore = () => {
            if (hasUndoManSelf) {
              undoManager.trackedOrigins.add(undoManager)
            }
            undoManager.doc.on('afterTransaction', undoManager.afterTransactionHandler)
            undoManager._observers = observers
          }
          viewRet.destroy()
        }
      }
    }
    return [
      ySyncPlugin(fragment),
      yUndoPluginInstance
    ]
  }
})
```

**Comment by DenysPoliarush** [comment]
It doesn't work for me, are there any other solutions to this problem?

**Comment by hamflx** [comment]
> It doesn't work for me, are there any other solutions to this problem?

You can try my solution: https://github.com/yjs/y-prosemirror/issues/114#issuecomment-1180235892

**Comment by Alecyrus** [comment]
@DenysPoliarush Downgrade the `y-prosemirror` to v1.0.20

**Comment by DenysPoliarush** [comment]
@Alecyrus It seems to have worked, I will test it, thanks

**Comment by codewitch** [comment]
> > It doesn't work for me, are there any other solutions to this problem?
> 
> You can try my solution: [yjs/y-prosemirror#114 (comment)](https://github.com/yjs/y-prosemirror/issues/114#issuecomment-1180235892)

hi @hamflx. Your solution works like a charm! Thank you. To make sure I understood your code correctly, could you confirm my understanding? It customizes the destroy logic so the yUndoManager gets restored on the view destroy life cycle? Would that create any memory leaks in the undomanager never getting destroyed?

**Comment by andrem0** [comment]
Why is this closed? The problem persists, and only a poor workaround of copying the original extension currently exists.

**Comment by svenadlung** [comment]
We released the fix in 2.0.0.
